ASoC: SOF: ipc4-pcm: Do not reset the ChainDMA if it has not been sta…

ujfalusi opened the pull request:
…rted

The ChainDMA operation differs from normal pipelines that it is only created when the stream started, in fact a PCM using CHainDMA has no pipelines, modules.
To reset a ChainDMA, it needs to be first initialized. When PulseAudio/PipeWire starts, they will probe the PCMs by opening them, check hw_params and then close the PCM without starting audio. Unconditionally resenting the CHainDMA can result the following error:

ipc tx      : 0xe040000|0x0: GLB_CHAIN_DMA
ipc tx reply: 0x2e000007|0x0: GLB_CHAIN_DMA
FW reported error: 7 - Unsupported operation requested
ipc error for msg 0xe040000|0x0
sof_pcm_stream_free: pcm_ops hw_free failed -22

The SOF_IPC4_PIPE_RESET state is only used during pcm_free. Skip sending the IPC message to reset the ChainDMA if it has not been initialized.

ujfalusi committed:
ASoC: SOF: ipc4-pcm: Do not reset the ChainDMA if it has not been allocated

The ChainDMA operation differs from normal pipelines that it is only
created when the stream started, in fact a PCM using ChainDMA has no
pipelines, modules.
To reset a ChainDMA, it needs to be first allocated in firmware. When
PulseAudio/PipeWire starts, they will probe the PCMs by opening them, check
hw_params and then close the PCM without starting audio.
Unconditionally resetting the ChainDMA can result the following error:

ipc tx      : 0xe040000|0x0: GLB_CHAIN_DMA
ipc tx reply: 0x2e000007|0x0: GLB_CHAIN_DMA
FW reported error: 7 - Unsupported operation requested
ipc error for msg 0xe040000|0x0
sof_pcm_stream_free: pcm_ops hw_free failed -22

Add a new chain_dma_allocated flag to sof_ipc4_pcm_stream_priv to store the
ChainDMA allocation state and use this flag to skip sending the reset if
the ChainDMA is not allocated.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -253,7 +253,7 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
 
 static int sof_ipc4_chain_dma_trigger(struct snd_sof_dev *sdev,
-				      int direction,
+				      struct snd_sof_pcm *spcm, int direction,
```
commit message:
resenting the CHainDMA

resetting the ChainDMA?

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -281,6 +281,14 @@ static int sof_ipc4_chain_dma_trigger(struct snd_sof_dev *sdev,
+		 * The ChainDMA can only be reset if it has been initialzied,
+		 * which is only done at stream start.
+		 */
+		if (!spcm->pending_stop[direction])
```
Yikes. 

This is re-using a state variable for a completely different purpose.
Are we sure this is the right variable? I already get a migraine to just think of the state machine that will use this.

Also this is only used for pcm_ops->platform_stop_during_hw_free, I don't think you can use it unconditionally.

ujfalusi replied:
IPC4 always uses this flag on trigger stop, which needs the stream to be started.
I was trying to avoid a new bool[2] for the ChainDMA initialized, but looks like I cannot.
The ChainDMA is a different construct than normal PCMs.

Let me revisit this next week.

plbossart replied:
also is this really an issue that needs to be fixed at the kernel level? Why not fix this at the firmware level? 

Or is this a problem with the reference firmware? I am not sure how we didn't detect this earlier if this was a problem with SOF proper.

ujfalusi replied:
@plbossart, yes, I have noticed it with the reference firmware (on UP2 after a kernel update).
We did not detect this as we stopped testing the reference firmware with Linux a long time ago and I think the postpone the stop to later in case of IPC4 came later.
This worked some time ago fine. What is not tested is not working (non ChainDMA on HDA with cAVS reference fw also broken).

ujfalusi commented:
Changes since v1:
- introduce a new flag `chain_dma_allocated` and use it to track the ChainDMA allocation state and to skip the reset when it would it invalid (the ChainDMA is not allocated)

It might be better to split the header and code change?

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -332,6 +332,7 @@ struct snd_sof_pcm_stream {
 	struct snd_soc_dapm_widget_list *list; /* list of connected DAPM widgets */
 	bool d0i3_compatible; /* DSP can be in D0I3 when this pcm is opened */
 	unsigned int dsp_max_burst_size_in_ms; /* The maximum size of the host DMA burst in ms */
+	bool chain_dma_allocated; /* indicates the ChainDMA allocation state */
```
not sure if this is feasible but could this boolean be moved in the IPC-specific 'private' structure? It's a bit odd to have a state variable for something that is only defined for IPC4.

ujfalusi replied:
@plbossart , the `void *private` is used solely for the ipc4 time tracking (delay stuff) in ipc4-pcm.c, and I just moved that struct locally to there. This flag is also used only in ipc4-pcm.c. It will take quite a bit of churn to make space for this flag behind of the `private`....

But unfortunately I agree with you :(

ujfalusi replied:
on the other hand the spcm->stream[substream->stream].pipeline_list is also ipc4 only, it just have a generic sounding name.

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -338,7 +343,12 @@ static int sof_ipc4_chain_dma_trigger(struct snd_sof_dev *sdev,
-	return sof_ipc_tx_message_no_reply(sdev->ipc, &msg, 0);
+	ret = sof_ipc_tx_message_no_reply(sdev->ipc, &msg, 0);
+	/* Update the ChainDMA allocation state */
+	if (!ret && spcm->stream[direction].chain_dma_allocated != allocate)
```
this check is redundant no? spcm->stream[direction].chain_dma_allocated != allocate

ujfalusi replied:
We can set it unconditionally, yes. Pause/resume will cycle between RUNNING/PAUSE, but that's fine.

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -253,14 +253,14 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
  */
 
 static int sof_ipc4_chain_dma_trigger(struct snd_sof_dev *sdev,
```
typo in commit message: :resenting -> resetting

ujfalusi replied:
Oh, thanks!

ujfalusi committed:
ASoC: SOF: ipc4-pcm: Use consistent name for snd_sof_pcm_stream pointer

Throughout the file the pointer for snd_sof_pcm_stream is named either
'stream' or (wrongly) 'spcm' which confuses the reader.

Use 'sps' for the pointer name as it is the most common name used in SOF
codebase.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-pcm: Use consistent name for sof_ipc4_timestamp_info pointer

The pointer to sof_ipc4_timestamp_info named most of the time as
'time_info' only to be named as 'stream_info' or 'info' in two function.

Use the consistent name of 'time_info' throughout the file.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-pcm: Introduce generic sof_ipc4_pcm_stream_priv

Using the sof_ipc4_timestamp_info struct directly as sps->private data
is too restrictive, add a new generic sof_ipc4_pcm_stream_priv struct
containing the time_info to allow new information to be stored in a
generic way.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v2:
- move the flag handling ipc4-pcm local by introducing a new generic sof_ipc4_pcm_stream_priv struct
- some cleanup to code to use consistent naming

plbossart reviewed (approved):
Looks nice and clean, thanks @ujfalusi for the update.
